src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs

```diff
@@ -35,28 +40,155 @@ public SwitchParameter Unique
         private bool _unique;
         #endregion
 
+        /// <summary>
+        /// This param specifies you only want the top N items returned.
+        /// </summary>
+        [Parameter(ParameterSetName="Default")]
+        [ValidateRange(1,int.MaxValue)]
+        public int Top { get; set; } = 0;
 
         /// <summary>
-        /// Remove duplicates.
+        /// This param specifies you only want the bottom N items returned.
         /// </summary>
-        private static void RemoveDuplicates(OrderByProperty orderByProperty)
+        [Parameter(ParameterSetName="Bottom", Mandatory=true)]
+        [ValidateRange(1,int.MaxValue)]
+        public int Bottom { get; set; } = 0;
+
+        /// <summary>
+        /// Moves unique entries to the front of the list.
+        /// </summary>
+        private void MoveUniqueEntriesToFront(List<OrderByPropertyEntry> sortedData, OrderByPropertyComparer comparer, out int uniqueCount)
         {
-            int current = 0, lookAhead;
-            OrderByPropertyEntry currentObj = orderByProperty.OrderMatrix[current];
-            while (current + 1 < orderByProperty.OrderMatrix.Count)
+            // If we have sorted data then we know we have at least one unique item
+            uniqueCount = sortedData.Count > 0 ? 1 : 0;
+            // Move the first of each unique entry to the front of the list
+            for (int uniqueItemIndex = 0, nextUniqueItemIndex = 1; uniqueItemIndex < sortedData.Count && uniqueCount != Top; uniqueItemIndex++, nextUniqueItemIndex++)
             {
-                lookAhead = current + 1;
-                OrderByPropertyEntry lookAheadObj = orderByProperty.OrderMatrix[lookAhead];
+                // Identify the index of the next unique item
+                while (nextUniqueItemIndex < sortedData.Count && comparer.Compare(sortedData[uniqueItemIndex], sortedData[nextUniqueItemIndex]) == 0)
+                {
+                    nextUniqueItemIndex++;
+                }
 
-                if (orderByProperty.Comparer.Compare(currentObj, lookAheadObj) == 0)
+                // If there are no more unique items, break
+                if (nextUniqueItemIndex == sortedData.Count)
                 {
-                    orderByProperty.OrderMatrix.RemoveAt(lookAhead);
+                    break;
                 }
-                else
+
+                // Move the next unique item forward and increment the unique item counter
+                sortedData[uniqueItemIndex + 1] = sortedData[nextUniqueItemIndex];
+                uniqueCount++;
+            }
+        }
+
+        /// <summary>
+        /// Sort unsorted OrderByPropertyEntry data using an indexed min-/max-heap sort
+        /// </summary>
+        private List<OrderByPropertyEntry> Heapify(List<OrderByPropertyEntry> dataToSort, OrderByPropertyComparer orderByPropertyComparer, out int heapCount)
```

Comment:
Yeah, I should have made that change when I switched to in-place sorting. It's done locally now.